src/context/values.js

```diff
@@ -0,0 +1,720 @@
+
+  /**
+   * Start the used value tracker if it hasn't started yet.
+   *
+   * @param {!ContextProp} prop
+   * @return {!UsedDef}
+   * @private
+   */
+  startUsed_(prop) {
+    const {key, deps} = prop;
+    const usedByKey = this.usedByKey_ || (this.usedByKey_ = new Map());
+    let used = usedByKey.get(key);
+    if (!used) {
+      devAssert(deps.length <= MAX_DEPS);
+      used = {
+        prop,
+        subscribers: [],
+        value: undefined,
+        pending: Pending.NOT_PENDING,
+        counter: 0,
+        depValues: deps.length > 0 ? deps.map(EMPTY_FUNC) : EMPTY_ARRAY,
+        parentValue: undefined,
+        parentValueNode: null,
+        // Schedule the value recalculation, optionally with the parent
+        // refresh.
+        ping: (refreshParent) => {
+          if (this.isConnected_()) {
+            const pending = refreshParent
+              ? Pending.PENDING_REFRESH_PARENT
+              : Pending.PENDING;
+            used.pending = Math.max(used.pending, pending);
+            this.checkUpdates_();
+          }
+        },
+        // Schedule the value recalculation due to the dependency change.
+        pingDep:
+          deps.length > 0
+            ? deps.map((dep, index) => (value) => {
+                used.depValues[index] = value;
+                used.ping();
+              })
+            : EMPTY_ARRAY,
+        // Schedule the value recalculation due to the parent value change.
+        pingParent: isRecursive(prop)
+          ? (parentValue) => {
+              used.parentValue = parentValue;
+              used.ping();
+            }
+          : null,
+      };
+      usedByKey.set(key, used);
+
+      // Subscribe to all deps.
+      deps.forEach((dep, index) => this.subscribe(dep, used.pingDep[index]));
+
+      // Schedule the first refresh.
+      used.ping();
+    }
+    return used;
+  }
+
+  /**
+   * Stop calculating the used value if there are no more subscribers left.
+   *
+   * @param {!UsedDef} used
+   * @private
+   */
+  stopUsed_(used) {
+    if (used.subscribers.length > 0) {
+      return;
+    }
+
+    const {prop, pingDep} = used;
+    const {key, deps} = prop;
+
+    this.usedByKey_.delete(key);
+
+    // Unsubscribe itself.
+    this.updateParentValueNode_(used, null);
+    if (deps.length > 0) {
+      deps.forEach((dep, index) => {
+        this.unsubscribe(dep, pingDep[index]);
+      });
+    }
+  }
+
+  /**
+   * Check if any properties awaiting recalculation. This method is always
+   * scheduled asynchronously and throttled.
+   *
+   * @private
+   */
+  checkUpdates_() {
+    if (!this.isConnected_()) {
+      // The node got disconnected between scheduling and handling.
+      return;
+    }
+
+    const usedByKey = this.usedByKey_;
+    if (!usedByKey) {
+      return;
+    }
+
+    usedByKey.forEach((used) => {
+      used.counter = 0;
+    });
+
+    let updated;
+    do {
+      updated = 0;
+      usedByKey.forEach((used) => {
+        if (used.pending) {
```

Comment:
Nit: We should explicitly check `!== NOT_PENDING`

Reply:
good idea.